mailpoet/lib/Subscribers/EngagementDataBackfiller.php

```diff
@@ -0,0 +1,200 @@
+    $subscriberIds = array_map(function (SubscriberEntity $subscriber) {
+      return $subscriber->getId();
+    }, $subscribers);
+
+    $clickData = $this->getClickDataForBatch($subscriberIds);
+    $openData = $this->getOpenDataForBatch($subscriberIds);
+    $sendingData = $this->getSendingDataForBatch($subscriberIds);
+    $purchaseData = $this->getPurchaseDataForBatch($subscriberIds);
+
+    foreach ($subscribers as $subscriber) {
+      if ($subscriber->getLastPurchaseAt() === null && isset($purchaseData[$subscriber->getId()]['last_purchase_at'])) {
+        $purchaseDate = new Carbon($purchaseData[$subscriber->getId()]['last_purchase_at']);
+        $subscriber->setLastPurchaseAt($purchaseDate);
+      }
+      if ($subscriber->getLastOpenAt() === null && isset($openData[$subscriber->getId()]['last_open_at'])) {
+        $openDate = new Carbon($openData[$subscriber->getId()]['last_open_at']);
+        $subscriber->setLastOpenAt($openDate);
+      }
+      if ($subscriber->getLastClickAt() === null && isset($clickData[$subscriber->getId()]['last_click_at'])) {
+        $clickDate = new Carbon($clickData[$subscriber->getId()]['last_click_at']);
+        $subscriber->setLastClickAt($clickDate);
+      }
+      if ($subscriber->getLastSendingAt() === null && isset($sendingData[$subscriber->getId()]['last_sending_at'])) {
+        $sendingDate = new Carbon($sendingData[$subscriber->getId()]['last_sending_at']);
+        $subscriber->setLastSendingAt($sendingDate);
+      }
+      if (is_int($subscriber->getId())) {
+        $this->lastProcessedSubscriberId = $subscriber->getId();
+      }
+    }
+
+    $this->entityManager->flush();
+  }
+
+  public function getClickDataForBatch(array $subscriberIds): array {
+    $subscribersTable = $this->filterHelper->getSubscribersTable();
+    $clicksTable = $this->filterHelper->getTableForEntity(StatisticsClickEntity::class);
+
+    $query = $this
+      ->entityManager
+      ->getConnection()
+      ->createQueryBuilder()
+      ->select("$subscribersTable.id, MAX(clicks.created_at) as last_click_at")
+      ->from($subscribersTable)
+      ->innerJoin($subscribersTable, $clicksTable, 'clicks', "$subscribersTable.id = clicks.subscriber_id")
+      ->andWhere("$subscribersTable.id IN (:subscriberIds)")
+      ->setParameter('subscriberIds', $subscriberIds, Connection::PARAM_INT_ARRAY)
+      ->groupBy("$subscribersTable.id");
+
+    $result = $query->execute();
+    if ($result instanceof Result) {
+      return $result->fetchAllAssociativeIndexed();
+    }
+    return [];
+  }
+
+  public function getPurchaseDataForBatch(array $subscriberIds): array {
+    if (!$this->wcHelper->isWooCommerceActive()) {
+      return [];
+    }
+
+    $subscribersTable = $this->filterHelper->getSubscribersTable();
+
+    $query = $this
+      ->entityManager
+      ->getConnection()
+      ->createQueryBuilder()
+      ->select("$subscribersTable.id, MAX(orderStats.date_created) as last_purchase_at")
```

Comment:
Is `orderStats` the correct alias? I am not too deep into the code but to me it looks like `wooFilterHelper` would use `customer` as default alias.

`wooFilterHelper->applyOrderStatusFilter` seems to return the alias, so it might make sense to use `applyOrderStatusFilter` and use it's return value instead of hardcoding `orderStats`

Reply:
Ah! Now I see `orderStats` is the default alias!

Reply:
I agree this is confusing, so I added a comment in 33c987db339cc2e5de0184e058f789afa569e834. I think the wooFilterHelper could be better in general, but I didn't feel it was necessary to change it as part of this work.